In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [3]:
dataset = pd.read_csv("Employee_Attrition_Dataset.csv")
dataset

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

In [4]:
dataset.drop(['EmployeeCount', 'EmployeeNumber','StandardHours'], axis = 1, inplace = True)

In [5]:
df1 = pd.get_dummies(dataset, dtype = int, drop_first = True)

#### Checking for Null Values

In [6]:
dataset.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [7]:
df1.columns

Index(['Age', 'DailyRate', 'DistanceFromHome', 'Education',
       'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel',
       'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager', 'Attrition_Yes',
       'BusinessTravel_Travel_Frequently', 'BusinessTravel_Travel_Rarely',
       'Department_Research & Development', 'Department_Sales',
       'EducationField_Life Sciences', 'EducationField_Marketing',
       'EducationField_Medical', 'EducationField_Other',
       'EducationField_Technical Degree', 'Gender_Male',
       'JobRole_Human Resources', 'JobRole_Laboratory Technician',
       'JobRole_Manager', 'JobRole_Manufacturing Director',
       'JobRole_Research Director', 'JobRole

# Univariate Analysis

In [8]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 32 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EnvironmentSatisfaction   1470 non-null   int64 
 9   Gender                    1470 non-null   object
 10  HourlyRate                1470 non-null   int64 
 11  JobInvolvement            1470 non-null   int64 
 12  JobLevel                  1470 non-null   int64 
 13  JobRole                   1470 non-null   object
 14  JobSatisfaction         

In [9]:
def qualQuan(dataset):
    quan = []
    qual = []
    for columnname in dataset.columns:
        if dataset[columnname].dtype == 'O':
            qual.append(columnname)
        else:
            quan.append(columnname)
    return quan, qual

In [10]:
quan, qual = qualQuan(dataset)

In [11]:
quan

['Age',
 'DailyRate',
 'DistanceFromHome',
 'Education',
 'EnvironmentSatisfaction',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobSatisfaction',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [12]:
qual

['Attrition',
 'BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'Over18',
 'OverTime']

In [13]:
df1['Attrition_Yes'].value_counts()

Attrition_Yes
0    1233
1     237
Name: count, dtype: int64

#### 1. How many Employees are highly job satisfied?

In [30]:
def freqtable(columnname, dataset):
    freqtable = pd.DataFrame(columns = ["Unique Values", "Frequency", "Relative Frequency", "Cumulative Frequency"])

    freqtable["Unique Values"] = dataset[columnname].value_counts().index
    freqtable["Frequency"] = dataset[columnname].value_counts().values
    freqtable["Relative Frequency"] = freqtable["Frequency"]/1470
    freqtable["Cumulative Frequency"] = dataset[columnname].cumsum()
    return freqtable

In [31]:
freqtable("JobSatisfaction", dataset)

Unique Values  Frequency  Relative Frequency  Cumulative Frequency
0              4        459            0.312245                     4
1              3        442            0.300680                     6
2              1        289            0.196599                     9
3              2        280            0.190476                    12

Out of 1470 Employees, 459 Employees are highly satisfied with their job. So they given the rating of 4.
569 Employees are not really satisfied with their job.

#### How many employees had a overall good worklife balance?

In [32]:
freqtable("WorkLifeBalance", dataset)

Unique Values  Frequency  Relative Frequency  Cumulative Frequency
0              3        893            0.607483                     1
1              2        344            0.234014                     4
2              4        153            0.104082                     7
3              1         80            0.054422                    10

Out of 1470 Employees, 153 had a good work life balance and 
893 had a moderate work life balance. 
424 employees are not really experience a good work life balance. 

#### How many Male and Female Employees?

In [33]:
dataset['Gender'].value_counts()

Gender
Male      882
Female    588
Name: count, dtype: int64

#### What is the average salary of a Male Employer?

In [34]:
male_avg = dataset[dataset["Gender"].str.contains('Male')]
Avg_male_salary = male_avg['MonthlyIncome'].mean()
print(Avg_male_salary)

6225.730158730159


In [35]:
male_avg = dataset[dataset["Gender"].str.contains('Male')]
Avg_male_salary = male_avg.groupby(['MonthlyIncome','MonthlyRate'])['PercentSalaryHike'].mean().reset_index()
print(Avg_male_salary.head(10))

   MonthlyIncome  MonthlyRate  PercentSalaryHike
0           1009        26999               11.0
1           1051        13493               15.0
2           1052        23384               22.0
3           1081        16019               13.0
4           1091        10642               17.0
5           1102         9241               22.0
6           1118         8040               14.0
7           1223        16901               22.0
8           1232        19281               14.0
9           1274         7152               13.0


So Average Salary of Male is 6225 with average monthly rate can assume like incentive of 26999. 
And Maximum Salary hike would be 22%